src/main/java/com/mongodb/kafka/connect/source/MongoCopyDataManager.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.kafka.connect.source;
+
+import static com.mongodb.kafka.connect.source.MongoSourceConfig.COLLECTION_CONFIG;
+import static com.mongodb.kafka.connect.source.MongoSourceConfig.COPY_EXISTING_MAX_THREADS_CONFIG;
+import static com.mongodb.kafka.connect.source.MongoSourceConfig.DATABASE_CONFIG;
+import static java.lang.String.format;
+import static java.util.Collections.singletonList;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+import org.apache.kafka.connect.errors.ConnectException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.bson.BsonDocument;
+import org.bson.conversions.Bson;
+
+import com.mongodb.MongoNamespace;
+import com.mongodb.client.MongoClient;
+
+
+/**
+ * Copy Data Manager
+ *
+ * <ol>
+ * <li>Gets all namespaces to copy. eg. A single collection, all collections in a database or all collections in all databases.</li>
+ * <li>For each namespace, submit to the executors an copy existing task.</li>
+ * <li>Each copy existing task, runs an aggregation pipeline on the namespace, to mimic an insert document and adds the results to the
+ *     queue.</li>
+ * <li>The poll method returns documents from the queue.</li>
+ * </ol>
+ */
+class MongoCopyDataManager implements AutoCloseable {
+    private static final Logger LOGGER = LoggerFactory.getLogger(MongoCopyDataManager.class);
+    private final AtomicBoolean closed = new AtomicBoolean();
+    private final AtomicBoolean errored = new AtomicBoolean();
+    private volatile String errorMessage;
+    private final AtomicInteger namespacesToCopy;
+    private final MongoSourceConfig sourceConfig;
+    private final MongoClient mongoClient;
+    private final ExecutorService executor;
+    private final ConcurrentLinkedQueue<BsonDocument> queue = new ConcurrentLinkedQueue<>();
```

Comment:
I think we can have that as a limitation of the copying process. Another queue type would just lead to an alternative error condition.

Reply:
I guess the alternative would be cursor timeouts (aggregate doesn't support noCursorTimeout flag afaik).  But I have a suspicion that cursor timeouts are going to be less likely than OOM, because the default timeout is fairly long (10 minutes iirc).  And blocking queues are a tried and true means for rate-limiting in producer/consumer algorithms so I'm not too concerned about introducing it here.

Perhaps we need to do some real world testing to see how this works when used in anger.

Reply:
Updated